edgelet/iotedge/src/check.rs

```diff
@@ -0,0 +1,749 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std;
+use std::borrow::Cow;
+use std::collections::HashSet;
+use std::ffi::{CStr, OsStr, OsString};
+use std::fs::File;
+use std::net::TcpStream;
+use std::path::PathBuf;
+use std::process::Command;
+
+#[cfg(unix)]
+use failure::Fail;
+use failure::{self, Context, ResultExt};
+use futures::future::FutureResult;
+use futures::IntoFuture;
+#[cfg(unix)]
+use libc;
+use regex::Regex;
+use serde_json;
+
+use edgelet_config::{Provisioning, Settings};
+use edgelet_core::{self, UrlExt};
+use edgelet_docker::DockerConfig;
+
+use error::{Error, ErrorKind};
+
+pub struct Check {
+    config_file: PathBuf,
+    steps_override: Option<Vec<String>>,
+    ntp_server: String,
+    latest_versions: super::LatestVersions,
+
+    // These optional fields are populated by the pre-checks
+    settings: Option<Settings<DockerConfig>>,
+    docker_host_arg: Option<String>,
+    iothub_hostname: Option<String>,
+}
+
+impl Check {
+    pub fn new(
+        config_file: PathBuf,
+        ntp_server: String,
+        steps_override: Option<Vec<String>>,
+        latest_versions: super::LatestVersions,
+    ) -> Self {
+        Check {
+            config_file,
+            ntp_server,
+            steps_override,
+            latest_versions,
+
+            settings: None,
+            docker_host_arg: None,
+            iothub_hostname: None,
+        }
+    }
+
+    fn execute_inner(&mut self) -> Result<(), Error> {
+        const PRE_CHECK_SECTION_NAME: &str = "pre-check";
+
+        // DEVNOTE: Keep the names of top-level steps in sync with help-text of the "check" subcommand in main.rs (except for "pre-check")
+        const CHECKS: &[(
+            &str,
+            &[(
+                &str,
+                fn(&mut Check) -> Result<Option<String>, failure::Error>,
+            )],
+        )] = &[
+            (
+                PRE_CHECK_SECTION_NAME,
+                &[
+                    ("config.yaml is well-formed", parse_settings),
```

Comment:
What is the format of this array of structures? Perhaps it would help to make this a type so that it is self documenting?

Reply:
Added comments on the type declaration.